Calculator App

Raupinger opened the pull request:
A calculator based on the Shunting-yard algorithm as described [here](https://en.wikipedia.org/wiki/Shunting-yard_algorithm). The basic functionality is in place but some work still needs to be done:
- [x] Second page of buttons for braces, pow etc.
- [x] Testing various edge cases
- [x] Haptic feedback for invalid operations
- [x] Clean up the UI
- [x] Make sure things like Precedence are handled correctly

this resolves #227

Raupinger commented:
Video

https://user-images.githubusercontent.com/29229062/119008299-56f20f00-b992-11eb-973b-1702db2fb84c.mp4

Raupinger commented:
Ok, I'm declaring this "ready for reviw" now. I cant find any bugs anymore and spent hours trying to reduce the memory usage. One obvious thing that could still be done is parsing the numbers while putting everything on the input stack. that would save n*(k)+31 bytes, where n is the number of digits a number has and k is the number of bytes a single digit on the stack takes up. in praxis it should be at the lower range of that. Overflows are not handled. The eval function has a lot of resemblance with Italian cuisine, so i wouldn't be surprised if there are still some errors in it. I couldn't test that the motor actually works but i don't expect any surprises there.

Raupinger commented:
[video](https://cdn.discordapp.com/attachments/622348681538043924/845857580189220904/VID_20210523_044523-50mB.mp4)

jonvmey commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+}
+
+void Calculator::eval() {
+  std::stack<char> input(std::deque<char>(0, ' '));
```
Just did a quick test and it looks like default constructing the three stack objects here saves 600 bytes on the binary size and unless I've missed something should have the exact same effect.

Raupinger replied:
you should be right

Raupinger replied:
changed it now

Dudemanjude commented:
I've found a bug that causes the watch to reset. If you just enter a period, and then hit equals, it resets. This happens no matter how many periods you add. For example, if you enter just "." and then hit equals, it will reset, and "...." and then equals will also reset. Another case of unwanted behavior happens when you put "1+.1" into the calculator, and hit equals. It says the answer is "2.1", while I would expect it to say "1.1". If you type ".2+.3", it results in "5", which is also unexpected.

The build that I'm running on my watch is a custom one that I compiled from your Calc branch, plus a custom Terminal watchface from a different repo. I'm a beginner when it comes to Git and compiling InfiniTime, so it's possible that my build is messed up, although everything else seems to work just fine.

Otherwise, the app is great, and I hope that once these bugs get sorted out, we can have a calculator built into one of the next versions of InfiniTime!

Raupinger commented:
Yeah, that sounds plausible. Currently there's no handling for periods that dont appear after some digits, so they just act as if nothing was there at all. The reset can be explained by trying to evaluate an empty expression. I should probably check for that and either interpret it as a zero or treat it as an invalid expression, vibrate and return

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+  int8_t sign = +1;
+  while (!input.empty()) {
+    
+    if (isdigit(input.top())) {
```
Yeah, someone else pointed that out, too. I personaly wouldnt consider .5 a legal number, but I have to add a check either to abort or to parse it as 0.5.

Raupinger replied:
I tried the Android calculator and KClac and they both read .5 as 0.5. Seems reasonable to follow that precedent.

Raupinger replied:
done now. This convention doesn't exist in German so English speakers rejoice!

Raupinger commented:
Thanks @jonvmey for all those reviews

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+        num += (numberStr[i + pointpos] - '0') / pow(10,i+1);
+      }
+  
+      NumNode* number = new NumNode();
```
I think there was some reason for using pointers. Ill have to take a look at that

Raupinger replied:
The reason was that you cant have the type of a field be abstract. I've now replaced the c-style pointers with shared_ptr (shared because there might be references in both the stack and individual nodes). Doing it this way appeases valgrind.

jonvmey replied:
Yes, I missed that these nodes get pushed onto the `output` stack.

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+        eval();
+      } else {
+        
+        text[position] = *buttonstr;
```
I was kind of aware of this but didn't want to deal with checking bounds. My rationale was that the buffer is sized to exactly match amount of characters that fit in the text field, so any input longer than that would be unusable anyways and "What maniac enters 30 character long calculations on their watch?". But you're right. Even if its unlikely, this has to be handled properly.

Raupinger replied:
fixed

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+      }
+  
+      NumNode* number = new NumNode();
+      number->value = num;
```
oh, i accidentally removed that when switching away from strtod.

Raupinger replied:
ok, that was pretty stupid. fixed it now

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,350 @@
+}
+
+void Calculator::eval() {
+  std::stack<char> input(std::deque<char>(0, ' '));
```
I don't see any tangible drawbacks from doing it this way.

jonvmey replied:
I agree the differences are pretty negligible. The main difference I could see is that iterating backwards by index instead of forwards is less intuitive and slightly more bugprone. Eg. If you iterate backwards and accidentally used an unsigned index variable instead of a signed one your loop never exits. There's nothing incorrect in your code that I can see, it's just something that I noticed.

Raupinger replied:
it doesn't really make a difference either way and i like the stack. I'll leave it like that

Izaic commented:
Is this just waiting to be merged or?

Raupinger commented:
Its should be ready. Its just been long enough that I would probably have to rebase it again

KoalaV2 commented:
Any updates on this?

Raupinger commented:
> Any updates on this?

It's basically done. At the time JF was to busy to take a look at this and its just kind of been sitting here since. Doesn't help that I've not been active on Infinitime for a while now. Rebasing it doesn't look to bad assuming it doesn't run into memory limits again.

fossdd commented:
i would really like that feature on InfiniTime

Alpha-Titiwu commented:
I have successfully tested this nice calculator by rebasing it on the latest develop branch. It would be great, if it could be merged into the official codebase.
@Raupinger could you please resolve the merge conflicts?

Vistaus commented:
This would be a handy feature indeed. 🙂

Raupinger commented:
Ok. I've rebased it all and fixed another bug found in testing. For now I have put the Calculator at the Meter apps spot in the app list and moved Meter to the settings list. It would be great if some of you could give this branch a try and tell me about any bugs i missed. If you don't have a devkit and are willing to take the relatively minimal risk you can download the dfu file from githubs automated build and flash it OTA.

Riksu9000 commented:
Inputting a large number and pressing equals returns -2147483648.
Inputting a large number that wraps on two lines has some weird behaviour at first where it takes at least three digits on the next line. This is because of an LVGL config. I don't know if there's an easy way to go around this, because this value probably shouldn't be changed.
https://github.com/InfiniTimeOrg/InfiniTime/blob/85a25302bfac215f9ec7b993f6a2c21a20ee223b/src/libs/lv_conf.h#L506
You can make the button matrix slightly larger by reducing the outer padding to zero like was done in the Tile screen. The inner padding can adjust the button density. https://github.com/InfiniTimeOrg/InfiniTime/blob/85a25302bfac215f9ec7b993f6a2c21a20ee223b/src/displayapp/screens/Tile.cpp#L94

```
lv_obj_set_style_local_pad_all(btnm1, LV_BTNMATRIX_PART_BG, LV_STATE_DEFAULT, 0);
lv_obj_set_style_local_pad_inner(btnm1, LV_BTNMATRIX_PART_BG, LV_STATE_DEFAULT, 10);
```

Raupinger commented:
I added a [file](https://github.com/Raupinger/Pinetime/blob/Calc/doc/Calculator.md) to the PR that explains some how to use the app.

Raupinger commented:
@Riksu9000
1. I'm aware of the overflow problem but I currently don't intend to do anything about it. People are rarely gonna work with values in that range on their wrist and I felt like checking for it would be a lot of work. That said i just looked it up and its pretty easy for addition, substraction and multiplication while pow() should report range errors. So this could and should probably be fixed
2. If anyone knows an easy fix for that, let me know, otherwise I don't think its that big a deal.
3. Good idea. Ill add that.

thx

Avamander commented:
> That said i just looked it up and its pretty easy for addition, substraction and multiplication while pow() should report range errors. So this could and should probably be fixed

I plan on disabling those errors. There's probably some less ideal but equally okay solution there.

Riksu9000 commented on src/displayapp/screens/Calculator.h:
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <array>
+#include "Screen.h"
+#include "../LittleVgl.h"
```
These includes should be non relative now. https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/coding-convention.md#coding-style

Raupinger replied:
Another relic from when most of this was written. Will fix.

Riksu9000 commented on src/displayapp/fonts/jetbrains_mono_bold_20.c:
```diff
@@ -22,1021 +22,1029 @@
 
 /*Store the image of the glyphs*/
 static LV_ATTRIBUTE_LARGE_CONST const uint8_t glyph_bitmap[] = {
-    /* U+0020 " " */
```
Please remake this file, so there aren't so many unnecessary changes.

Riksu9000 commented on src/displayapp/DisplayApp.cpp:
```diff
@@ -26,8 +26,10 @@
 #include "displayapp/screens/SystemInfo.h"
 #include "displayapp/screens/Tile.h"
 #include "displayapp/screens/Twos.h"
+#include "displayapp/screens/Calculator.h"
 #include "displayapp/screens/FlashLight.h"
 #include "displayapp/screens/BatteryInfo.h"
+#include "displayapp/screens/Calculator.h"
```
This is included twice.

Riksu9000 commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,374 @@
+  lv_obj_set_style_local_pad_all(buttonMatrix, LV_BTNMATRIX_PART_BG, LV_STATE_DEFAULT, 0);
+  buttonMatrix->user_data = this;
+  lv_obj_set_event_cb(buttonMatrix, eventHandler);
+  running = true;
```
The running variable doesn't need to be initialized or defined in the header, because it's already done in Screen.h

Riksu9000 reviewed (commented):
Entering just "+" and pressing equals crashes the firmware.
Swiping left and right might also input characters. I'll look into this more as well.
We could change the "+" to something else on swipe too. Maybe pi?
I think it's very likely that many people will encounter the overflow problem by just playing around with it, so I think we should do something about it.
Maybe the results should be on a separate line, so it doesn't erase the input. I think it's unlikely that people perform calculations that don't fit on a single line on a watch.
Please clang-format the Calculator.* files and fix the compiler warnings.

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,374 @@
+#pragma GCC push_options
+#pragma GCC optimize ("Os")
```
I started writing the calculator before that was added. You're right that it should be removed

Raupinger commented:
Ok, i've fixed the crash and added over/underflow checks. The code for the lather is neither particularly pretty or fast, but it works.

medeyko commented:
I'd propose to introduce % (remainder) operator, with button on the position of + on the second buttons set. It is often implemented in physical calculators.
Also it seems to be the right thing to introduce sin, cos, tan, arcsin, arccos, arctan, log, ln, exp, 1/x, n!, sqr, sqrt (maybe on a third buttons set) - these functions look cheap to implement, but turn the calculator into a scientific one, that may be used in more circumstances.
Still, of course, the calculator would be handy anyway.

Izaic commented:
> I'd propose to introduce % (remainder) operator, with button on the position of + on the second buttons set. It is often implemented in physical calculators. Also it seems to be the right thing to introduce sin, cos, tan, arcsin, arccos, arctan, log, ln, exp, 1/x, n!, sqr, sqrt (maybe on a third buttons set) - these functions look cheap to implement, but turn the calculator into a scientific one, that may be used in more circumstances. Still, of course, the calculator would be handy anyway.

That would be way beyond the scope of this calculator, it's just supposed to be a simple calc for doing quick math on your wrist in a pinch.

medeyko commented:
> That would be way beyond the scope of this calculator, it's just supposed to be a simple calc for doing quick math on your wrist in a pinch. 
I personally time to time calculate sqrt or sine exactly as a "quick math". Of course, I have a smart phone, but still, using watch may be more covenient. Considering that the option is cheap to implement, I don't see, why not.

>the calculator app for BangleJS added all of those features that you've mentioned, and it means that it's really hard to accurately hit buttons on the calculator as a result because they are now so tiny.
I don't mean to make the buttons any smaller, there's no any need to shrink them. I listed 12 functions - it's 4 by 3 rectangle. So they may be introduced on a secondary buttons screen, activated by swipe (either replacing numbers on current secondary buttons screen or making second secondary screen activated by the opposite swipe).

trman commented:
> I personally time to time calculate sqrt or sine exactly as a "quick math". Of course, I have a smart phone, but still, using a watch may be more covenient. Considering that the feature is cheap to implement, I don't see, why not.

it may be not cheap to implement @medeyko  since it may need to include a math library for the other function , and since infintime is tight in space ,  it would make the pr unmergable because the since : 
math lib + pr + firmware > firmware max flash  size  .

it's better the keep it simple first , in order to make the merge easier for @JF002  .

medeyko commented:
I absolutely agree that thing that make the merge difficult should be avoided at this stage!

However, I still disagree that my proposal falls under this category:
math.h/cmath ALREADY #included in the calculator app. It also contains sine/cosine support for Cortex M4F.

As for the pow function, it is already used in the calculator app. Moreover, its documentation suggests to use x^(1/2) for calculation of a square root.

Factorial is a quite simple function. So no ant additional math lib is required... 

Therefore I still think that my proposal is cheap to implement. But ok, I would be happy to have the calculator app even in its current state.

Avamander commented:
> math.h/cmath ALREADY #included in the calculator app. It also contains sine/cosine support for Cortex M4F.

Compiler usually optimizes out functions that aren't used.

> Moreover, its documentation suggests to use x^(1/2) for calculation of a square root.

This suggestion should not be followed before one verifies it's actually better.


In the end, the scope of the PR should be kept as-is, further feature requests should be placed in an appropriate issue.

jonvmey commented:
For reference #476 was done in order to **remove** usage of `cmath`'s sine/cosine functions as that saved ~7kB of binary size. Accurate trig functions (ie. beyond the LUT now used for the analog watchface) are relatively expensive size-wise.

Raupinger commented:
> - [ ] Swiping left and right might also input characters. I'll look into this more as well.

The same sometimes happens on the app select screen. There's probably some improvement to Infinitimes gesture recognition to be made in general here
> - [ ] We could change the "+" to something else on swipe too. Maybe pi?

I like the idea of adding a modulo operator
> - [ ] Maybe the results should be on a separate line, so it doesn't erase the input. I think it's unlikely that people perform calculations that don't fit on a single line on a watch.

That would have the added benefit of being able to display error messages instead of just vibrating. What would entering a new calculation look like with that?

Raupinger commented:
> - [] I think it's very likely that many people will encounter the overflow problem by just playing around with it, so I think we should do something about it.

That fix involved a fair bit of trial and error. Id to see someone take a shot at breaking it again. The last overflow check when converting the float to an inte has the potential to hide cases where an overflow happens during calculation and isn't caught by the checks there.

Raupinger commented:
@medeyko btw. the compiler optimizing away unused library function is also the reason for the somewhat awkward float to string conversion seen in the calculator. Not using the library functions for that saves 6k in the binary.

Raupinger commented on src/displayapp/screens/Calculator.cpp:
```diff
@@ -0,0 +1,374 @@
+}
+
+bool Calculator::OnTouchEvent(Pinetime::Applications::TouchEvents event) {
+  if (event == Pinetime::Applications::TouchEvents::LongTap) {
+    position = 0;
+    lv_label_set_text(result, "0");
+    return true;
+  }
```
Unless we want long presses on other keys to mean something i believe it's easier to use this way. Ofc it's also easier to trigger by accident if you think that's a problem.

medeyko commented:
Thanks, I agree, that 6k and 7k of memory is important. I thought that Cortex m4f contains hardware sine, sorry.

However, %, exp, 1/x, n!, sqr, sqrt and constants e and pi wouldn't require the memory sacrifices :)

Ok, no persuading anymore, just a small notice for @Avamander 
It seems that you didn't understand which documentation I meant, because you write:
> This suggestion should not be followed before one verifies it's actually better.
But currently there's no alternatives at all, there's nothing to verify, as far as I see.

I meant the tip here:
https://github.com/Raupinger/Pinetime/blob/Calc/doc/Calculator.md

Will wait for the merge!

NeroBurner commented:
rebased the Calc branch on current `develop`: https://github.com/NeroBurner/InfiniTime/tree/Calc_rebase

added the `lv_simulation` branch on top: https://github.com/NeroBurner/InfiniTime/tree/lv_simulation_Calc

Nice for local hacking:
![Screenshot_20220226_220416](https://user-images.githubusercontent.com/9076163/155858992-56affc26-f1b0-4689-ba26-c65caf96c24a.png)

Raupinger commented:
Just a quick heads up: Ive had some incorrect results (like 0.3*0.3) and am still not super confident about my overflow checks. I would like to pull out the eval function and write some tests for it before this gets realeased.

Vistaus commented:
Too bad this didn't get merged yet. I like that the main devs are active, but the main feature in 1.9.0 is a terminal watchface, really? I mean: it looks great and geeky, so I really dig it! :) But if I had to choose, I would've liked to see something that's actually useful, like this calculator app.

medeyko commented:
> Just a quick heads up: Ive had some incorrect results (like 0.3*0.3) and am still not super confident about my overflow checks. I would like to pull out the eval function and write some tests for it before this gets realeased.

Had you an opportunity to raise your confidence in overflow checks? If not, could you please write more details?

JF002 commented:
> Too bad this didn't get merged yet. I like that the main devs are active, but the main feature in 1.9.0 is a terminal watchface, really? I mean: it looks great and geeky, so I really dig it! :) But if I had to choose, I would've liked to see something that's actually useful, like this calculator app.

Please note that there are a lot of reasons why any specific PR is not merged yet : some are not finished or do not meet the project goals, priorities and/or code quality, they need time and attention from developers to review it and maintain it, they need too much memory and cannot be merged until we can find a way to free some ram/flash memory space,... 

As you can see, there are currently ~100 open PR, and I'm really sorry I cannot give them the attention they deserve, but I (along with other core developers) work on this project on our free time, and we cannot possibly review and merge that much PR as soon as they are opened.
Also, remember that the PineTime is a very constrained hardware (in terms of processing power and memory). In the end, it'll be impossible to merge each and every features without hitting processing or memory limitations, so choices have to be made, unfortunately.

I'm sorry you don't find the terminal watchface as amazing as I do, but I'm pretty sure that If we merged this one instead, other people would have said that they don't need a calculator, and that they are sad the new watchface is not released yet.
Also, note that the terminal watchface is not the only addition in 1.9 : [49 PR](https://github.com/InfiniTimeOrg/InfiniTime/milestone/9) and [100 commits](https://github.com/InfiniTimeOrg/InfiniTime/compare/1.8.0...1.9.0) were merged since 1.8. Maintaining software do not only consist in adding feature visible to the end user. It also consist in fixing bugs, improving the code that is running under the hood, improving performances and maintainability of the code,...

Now, regarding this PR, the first question I'll ask before going any further is : what is the added memory usage (flash and ram ) of this PR ?

Danimations commented:
Just writing to express my support for this feature. I would definitely use it once it's available.

NeroBurner commented:
rebased the Calc branch on current `develop`: https://github.com/NeroBurner/InfiniTime/tree/Calc_rebase

Additionally I refactored the PR into three single commits to make rebases easier when the fonts were changed by other PRs:
- fonts/README: add calculator symbol to font generation documentation
- fonts: regenerate jetbrains_mono_bold_20 with calc symbol
- Add Calculator screen as new App (the changes from @Raupinger )

I propose to replace the current branch content with the new `Calc_rebase` commits. To do that you can do the following steps (assuming `origin` is `git@github.com:Raupinger/Pinetime.git`)

```sh
git remote add neroburner https://github.com/NeroBurner/InfiniTime.git
git fetch neroburner Calc_rebase
git switch Calc
git reset --hard neroburner/Calc_rebase
git push --force-with-lease origin Calc
```

Raupinger commented:
Thanks!

rev22 commented:
I tried to reduce the firmware size and ram usage by using statically allocated stacks, and removing the intermediate tree data structure, without changing the algorithm.
Some issues were solved, for example with displaying numbers like `0.09`, and with expressions like `-(2^32)+1`
Feel free to cherry pick the commit: https://github.com/rev22/InfiniTime/commit/59622e1e1725fe84b93f9f66660a08375572ad48 (code not very polished, still testing)
Firmware size changed from 13164 to 8660 bytes in my test builds with `gcc-arm-11.2-2022.02` and `-O2 -Os`.  I suppose part of it is the lv_btnmatrix component

ItzSwirlz commented:
Any status on this besides a potential 1.10 rebase if needed? I'd love to see this merged

MaWalla commented:
I merged the rebase from @NeroBurner into the current develop state and resolved all conflicts. Hope I did it right! :smile: 

https://github.com/MaWalla/InfiniTime/tree/Calc_rebase

Raupinger commented:
Neat. Looking forward to seeing the improvements you people made in an upcoming release.

omginput commented:
Could someone please tell why this hasn't been merged yet?

JustScott commented:
I updated the code to work with v1.14, its available on my branch: https://github.com/JustScott/InfiniTime/tree/calculator_app.

While it's working in InfiniSim, I get the following error attempting to build the mcuboot-app directly: 
```
/Pinetime/gcc-arm-11.2-2022.02-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/11.2.1/../../../../arm-none-eabi/bin/ld: region RAM overflowed with stack Memory region
```
I'm assuming this means the `Calculator.cpp` code needs some memory optimizations before the app can run on the Pinetime's hardware...which is a bit above my skill level at the moment, so if anyone would like to try this out feel free to use my branch!

julianfoad commented:
# Memory Optimisations

@JustScott re. "needs some memory optimizations", this commit solves it for me: https://lab.trax.im/gentle/infinitime/-/commit/eebd027eecb2a4c756d02efecd11f96fc914d95a

> Calculator: move 160 bytes from RAM to flash
> 
> Making the 'button matrix' arrays 'const' saves 160 bytes of non-stack
> RAM, which in this context is a lot, and makes it fit (for me) in a
> real-device build along with my choice of other apps.

The essence of the change is:

```diff
@@ -124,3 +124,3 @@ Calculator::~Calculator() {
 
-static const char* buttonMap1[] = {
+static const char* const buttonMap1[] = {
   "7", "8", "9", "/", "\n",
@@ -131,3 +131,3 @@ static const char* buttonMap1[] = {
 
-static const char* buttonMap2[] = {
+static const char* const buttonMap2[] = {
   "7", "8", "9", "(", "\n",
```

It also includes three `const_cast`s which are needed to overcome what appears to be an oversight in LVGL's const-correctness, which I have verified to be a safe casting.

You can pull that specific commit from my repo into your `calculator_app` branch like this (as I choose not to do my work in Microsoft Github):

```sh
git pull https://lab.trax.im/gentle/infinitime/ eebd027eecb2a4c756d02efecd11f96fc914d95a
```

or pull my `calculator_app` branch (which is based on yours) like this:

```sh
git pull https://lab.trax.im/gentle/infinitime/ calculator_app
```

which is currently just that single commit but I might add more.

Let me know if it works for you, if you try it. Thanks.

tituscmd commented:
Anything holding this back from being merged?

julianfoad commented:
@tituscmd yes of course there is. please read the comments especially https://github.com/InfiniTimeOrg/InfiniTime/pull/375#issuecomment-1100837137

tituscmd commented:
That comment being 3 years old, I thought I might as well check up on it again. My bad!

mark9064 commented:
There's also #1483 to consider

L3br4nd commented:
Yes Stock 1.16. But I did not install the resources. Maybe that causes the issue then.
